base/bitarray.jl

```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
 
 ## map over bitarrays ##
 
-function map!(f::Callable, A::Union(StridedArray,BitArray))
-    for i = 1:length(A)
-        A[i] = f(A[i])
-    end
-    return A
-end
-
-function map!(f::Callable, dest::Union(StridedArray,BitArray), A::Union(StridedArray,BitArray))
-    for i = 1:length(A)
-        dest[i] = f(A[i])
-    end
-    return dest
-end
-
-function map!(f::Callable, dest::Union(StridedArray,BitArray), A::Union(StridedArray,BitArray), B::Union(StridedArray,BitArray))
-    for i = 1:length(A)
-        dest[i] = f(A[i], B[i])
-    end
-    return dest
-end
-
-function map!(f::Callable, dest::Union(StridedArray,BitArray), A::Union(StridedArray,BitArray), B::Number)
-    for i = 1:length(A)
-        dest[i] = f(A[i], B)
-    end
-    return dest
-end
-
-function map!(f::Callable, dest::Union(StridedArray,BitArray), A::Number, B::Union(StridedArray,BitArray))
-    for i = 1:length(B)
-        dest[i] = f(A, B[i])
-    end
-    return dest
-end
-
-function map!(f::Callable, dest::Union(StridedArray,BitArray), As::Union(StridedArray,BitArray)...)
-    n = length(As[1])
-    i = 1
-    ith = a->a[i]
-    for i = 1:n
-        dest[i] = f(map(ith, As)...)
-    end
-    return dest
+# Specializing map is even more important for bitarrays than it is for generic
+# arrays since there can be a 64x speedup by working at the level of Int64
+# instead of looping bit-by-bit.
+
+# Note that there are 16 possible pure two-argument logical functions,
+# of which six are trivial and two don't exist as a single function in Base:
+##############################################################################
+##  p = TTFF                          ##  p = TTFF                          ##
+##  q = TFTF    function  name        ##  q = TFTF    function  name        ##
+##  --------------------------------  ##  --------------------------------- ##
+##      TTTT    (true)    -           ##      FFFF    (false)   -           ##
+##      TTTF    |, max    or          ##      FFFT    ???       (nor)       ##
+##      TTFT    >=, ^     ???         ##      FFTF    <         ???         ##
+##      TTFF    (A)       -           ##      FFTT    (~A)      -           ##
+##      TFTT    <=        implies     ##      FTFF    >         notimplies  ##
+##      TFTF    (B)       -           ##      FTFT    (~B)      -           ##
+##      TFFT    ==        xnor        ##      FTTF    $, !=     xor         ##
+##      TFFF    &, *, min and         ##      FTTT    ???       (nand)      ##
+##############################################################################
+
+abstract Func{N}
+
+immutable NotFun <: Func{1} end
+call(::NotFun, x) = ~x
+
+# Use IdFun, AndFun, and OrFun from reduce.jl
+
+immutable POrNotQFun <: Func{2} end
+call(::POrNotQFun, x, y) = x | ~y
```

Comment:
If we defined `call(::Type{POrNotFun}, x, y)` rather than `call(::POrNotQFun, x, y)`, and returned `POrNotFun` rather than `POrNotFun()` from `specialized_binary`, and of course analogous definitions for the other functions, then we could use `Func{1}`/`Func{2}` in the type signature of `map`, which I think would be safer/clearer.